File: epi_judge_python/is_string_palindromic_punctuation.py
from test_framework import generic_test
import string


def is_palindrome(s: str) -> bool:
    fr = 0
    ba = len(s) - 1
    while fr < ba:
        while fr < len(s) and (
            s[fr] in string.punctuation or s[fr] in string.whitespace
        ):
            fr += 1
        while ba >= 0 and (s[ba] in string.punctuation or s[ba] in string.whitespace):
            ba -= 1
        if fr < ba and s[fr].lower() != s[ba].lower():
            return False
        fr += 1
        ba -= 1
    return True


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "is_string_palindromic_punctuation.py",
            "is_string_palindromic_punctuation.tsv",
            is_palindrome,
        )
    )

File: epi_judge_cpp/is_string_palindromic_punctuation.cc
#include <_ctype.h>
#include <iterator>
#include <string>

#include "test_framework/generic_test.h"
using std::string;
bool IsPalindrome(const string& s) {
  if (s.empty() || s.size() == 1) {
    return true;
  }
  auto fi = s.begin();
  auto bi = s.rbegin();
  while (std::distance(s.begin(), fi) + std::distance(s.rbegin(), bi) <= s.size()) {
    while (!isalnum(*fi)) {
      fi++;
    }
    while (!isalnum(*bi)) {
      bi++;
    }
    if (toupper(*fi) != toupper(*bi)) {
      return false;
    }
    fi++;
    bi++;
  }
  return true;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"s"};
  return GenericTestMain(args, "is_string_palindromic_punctuation.cc",
                         "is_string_palindromic_punctuation.tsv", &IsPalindrome,
                         DefaultComparator{}, param_names);
}
